### ISSUE TITLE
`if` + `then` rejected with significant indentation off

### ISSUE THREAD

**Issue by kubukoz** [scalameta/scalafmt#3941]
## Configuration (required) ##

```
version = 3.8.1
runner.dialect = scala3
runner.dialectOverride.allowSignificantIndentation = false
```

## Command-line parameters (required) ##

When I run scalafmt via CLI like this: `scalafmt --check foo.scala`

## Steps

Given code like this:

```scala
object hello {
  val x = if true then true else false
}
```

## Problem

Scalafmt fails:

```scala
org.scalafmt.dynamic.exceptions.PositionExceptionImpl: /private/var/folders/f_/cbwjhbrs2jjg59qn23y6nhvm0000gn/T/tmp.87QcAG4Rnd/foo.scala:2: error: [dialect scala3 [with overrides]] `then` expected but `else` found
  val x = if true then true else false
                            ^: /private/var/folders/f_/cbwjhbrs2jjg59qn23y6nhvm0000gn/T/tmp.87QcAG4Rnd/foo.scala
error: ParseError=2
```

## Expectation

I would like the formatting to succeed.

## Workaround

Not using `if then` with `allowSignificantIndentation=false`.

## Notes

I don't think `allowSignificantIndentation` should prevent the use if `if then` with single expressions (scalac doesn't complain in `-no-indent` mode).

**Comment by kitbellew** [comment]
@kubukoz why did you need to disable significant indentation?

**Comment by kubukoz** [comment]
> @kubukoz why did you need to disable significant indentation?

I didn't need to, I wanted to :) my team chose to keep using braces, and by having Scalafmt know about this, we let it fix the indents on e.g. copy-pasted code.

**Comment by tgodzik** [comment]
This is expected `runner.dialectOverride.allowSignificantIndentation` is what enables the parser to actually parse significant indentation and `if-then` is a part of that. We would need to make a separate dialect option for that and I don't believe it's really needed.

Wouldn't it make sense to just disable it via compiler options?

**Comment by tgodzik** [comment]
`scalacOptions += "-no-indent"` works the way you want

**Comment by kitbellew** [comment]
the question is: if you keep indentation enabled, does scalameta parse incorrectly? doc says that code is not simply parsed as if there had been no significant indentation but may not be written with multiple indented expressions without explicit braces.

or, perhaps, even if scalameta parses correctly, does scalafmt format in a way that the compiler doesn't like? that is, if RedundantBraces is not used. 

if the answers are both no, then why disable the flag?

**Comment by kubukoz** [comment]
> `scalacOptions += "-no-indent"` works the way you want

Yes, I'm using that - I have no trouble with the compiler's side of things.

The problem is, if I remove the `allowSignificantIndentation = false`, part, then scalafmt doesn't by itself fix the indentation, so I have to rely on compiler warnings to do so:

<img width="985" alt="image" src="https://github.com/scalameta/scalafmt/assets/894884/8c77bae7-3ee5-49b1-85dc-e814b22d7c38">

`RedundantBraces` isn't enabled here, but I do use it in the real project. With that option, we get a completely different behavior, fixing of which takes an additional step (removing another indent after you still see the warning). It may be troublesome to do it every time you copy-paste / move a chunk of code.

<img width="887" alt="image" src="https://github.com/scalameta/scalafmt/assets/894884/a5d727da-934e-4b02-a1b6-accfbe72e4ec">

---

to sum up, I'm looking for something like `-no-indent` but for Scalafmt. `if then` is [the recommended conditional syntax](https://docs.scala-lang.org/scala3/book/control-structures.html#the-ifthenelse-construct) in the docs, so I'd like to keep using that.

Side note, when braceless syntax was being introduced we were told it'd still be possible to write code _with_ braces - that is all I'm trying to achieve here :)

**Comment by kubukoz** [comment]
Admittedly, Metals/vscode has gotten a bit better at moving lines up and down, so maybe the issue isn't significant enough and we can remove the flag now. That said, it'd be a bit disappointing/surprising if the formatter is able to change code behavior without an opt-out key (for non-indent users).

**Comment by tgodzik** [comment]
Ach ok, I see the exact scenario, I couldn't figure out the problem previously. There is no way to do that now, we would need to change the dialect to separate `then` etc. options in scalameta

**Comment by kitbellew** [comment]
@kubukoz when you say "fix the indentation", what exactly do you mean? if your indentation was incorrect to begin with, did you expect the formatter to fix it (and how)? if that indentation was in fact correct, then what happened?

what is the original code you have (before scalafmt formatting)? and what's the output?

**Comment by kubukoz** [comment]
> what is the original code you have (before scalafmt formatting)? and what's the output?

Original:

```scala
//> using scala 3.4.1
//> using option -no-indent

object foo {
  def greet() = println("hello")
  def sayGoodbye() = println("goodbye")
  def openDoor() = println("door opened")

  def demo() =
    if (true) greet()
    else
      sayGoodbye()
      openDoor()

}
```

Formatted with:

```
version = 3.8.1
runner.dialect = scala3
```

it's the same as the input.

Expectation (also what happens in `dialect = scala213`):

```scala
//> using scala 3.4.1
//> using option -no-indent

object foo {
  def greet() = println("hello")
  def sayGoodbye() = println("goodbye")
  def openDoor() = println("door opened")

  def demo() =
    if (true) greet()
    else
      sayGoodbye()
  openDoor()

}
```

> if your indentation was incorrect to begin with, did you expect the formatter to fix it 

It's correct in Scala 2 style: AFAIK nesting code too far in S2 isn't considered incorrect, and the formatter normally turns cases like that into consistently/correctly indented code.

**Comment by kitbellew** [comment]
according to 2.2 (https://docs.scala-lang.org/scala3/reference/other-new-features/indentation.html), your original code is invalid under `-no-indent`. does it compile with scala 3 compiler?

**Comment by kubukoz** [comment]
It does, with warnings.

**Comment by kitbellew** [comment]
fixed in #4050
